### DISCUSSION TITLE
Performance issues with useFieldArray

### DISCUSSION THREAD

**Discussion by wdfinch** [react-hook-form/react-hook-form#4632] [Q&A]
I'm having some performance issues with large field arrays. Specifically, I have complex field array elements (many fields on each element) and I need to mount every one of them in order for `useFieldArray` to work correctly. I am using `display: none`, but just the operation to update the dom is very expensive. What I would like to do is just mount one or a few at a time, however, this will cause issues as the array elements won't be connected.

@bluebill1049 is there anything I can do here to improve the performance? 

One idea I had was still connecting the element but not mounting all the inputs. So something like:

```
{fields.map((field, index) => (
  <div key={field.id}>
     {hideElement && (
       // the inputs
     )
  </div>
))}
```

I feel like this still might cause issues though. I'm just not sure what to do when my forms have huge field arrays and each element has many inputs. Trying to mount everything becomes unusably slow.

**Comment by bluebill1049** [comment]
I think the problem is rendering all those external components, useFieldArrry doesn't really do much in terms of thread work. may I ask how big is your field array?

**Reply by wdfinch** [reply]
I have about 100 elements but each of those elements contains numerous inputs. The issue is for enterprise users, some of their forms are massive.

**Reply by bluebill1049** [reply]
I would try to avoid re-render them, is the initial render laggy? or could you prevent from re-render or even remount the entire field array?

**Reply by wdfinch** [reply]
Re-render is fine. It's the initial render that's taking a long time. So I guess not the end of the world, but the user can be waiting a long time for that first load and it will keep getting worse as the form grows.

**Reply by bluebill1049** [reply]
I see I would try to use more native inputs. Controlled components from the external library may have additional logic.

**Reply by wdfinch** [reply]
Yes unfortunately I'm using inputs like react select and even some custom ones I made and I can't really move away from these.

**Reply by bluebill1049** [reply]
React Select is definitely not cheap to run for 100 plus times.

**Reply by wdfinch** [reply]
Yes, but I need the features it offers. I guess would the work you did on #4359 give me any options for not mounting any parts of a field array? I could always just test stuff in code sandbox, but I don't want to go against the design principles in case something changes down the road. So for example could I get away with not mounting elements, inputs of elements, etc., and still safely get values via the shallow merge of default values + mutate form data even if all the field array elements are not mounted? 

Otherwise, I don't really see a solution here aside from not using my input components.
